src/relying_party/mod.rs

```
@@ -0,0 +1,224 @@
+use crate::{
+    error::OAuthTSLError,
+    tokens::{
+        referenced_token::{ReferencedToken, ReferencedTokenClaims},
+        status_list_token::{StatusListToken, StatusListTokenClaims, StatusListTyp},
+    },
+};
+use flate2::read::GzDecoder;
+use jsonwebtoken::{decode, decode_header, DecodingKey, Validation};
+use std::io::Read;
+
+/// The media types defined for status list tokens.
+#[derive(Debug)]
+pub enum StatusListTokenResponseType {
+    Jwt,
+    Cwt,
+}
+
+impl StatusListTokenResponseType {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            StatusListTokenResponseType::Jwt => "application/statuslist+jwt",
+            StatusListTokenResponseType::Cwt => "application/statuslist+cwt",
+        }
+    }
+}
+
+impl TryFrom<&str> for StatusListTokenResponseType {
+    type Error = OAuthTSLError;
+
+    fn try_from(value: &str) -> Result<Self, OAuthTSLError> {
+        match value {
+            "application/statuslist+jwt" => Ok(StatusListTokenResponseType::Jwt),
+            "application/statuslist+cwt" => Ok(StatusListTokenResponseType::Cwt),
+            _ => Err(OAuthTSLError::InvalidContentType),
+        }
+    }
+}
+
+/// Decrypt and validate a referenced token jwt
+pub fn decrypt_referenced_token_jwt(
+    token_jwt: &str,
+    decoding_key: DecodingKey,
+) -> Result<ReferencedToken, OAuthTSLError> {
+    let header = decode_header(token_jwt)?;
+    if header.typ != Some(StatusListTyp::Jwt.as_string()) {
+        return Err(OAuthTSLError::InvalidHeaderTypeClaim(format!(
+            "{:?}",
+            header.typ
+        )));
```

nit: you're formatting part of the error here, and then some more formatting here:
```rust
    #[error("Invalid type claim (`typ`) in jwt header: {0}")]
    InvalidHeaderTypeClaim(String),
```

you could choose to do all formatting using `thiserror` like so:
```rust
    #[error("Invalid type claim (`typ`) in jwt header: {0:?}")]
    InvalidHeaderTypeClaim(Option<String>),
```

so that you can return the error like this:
```rust
return Err(OAuthTSLError::InvalidHeaderTypeClaim(header.typ));
```

---

fixed